OpenSim/Simulation/Test/testStatesTrajectory.cpp

```diff
@@ -508,6 +508,386 @@ void testCopying() {
     }
 }
 
+void testAccessByTime() {
+    Model model("arm26.osim");
+    auto& state = model.initSystem();
+
+    // Make a trajectory with different times.
+    StatesTrajectory states;
+    state.setTime(0.5);
+    states.append(state); // 0
+    state.setTime(1.3);
+    states.append(state); // 1
+    state.setTime(3.5);
+    states.append(state); // 2
+    state.setTime(4.1);
+    states.append(state); // 3
+    
+    // Nearest before.
+    SimTK_TEST(states.findIndexNearestBefore(0.5) == 0);
+    SimTK_TEST(states.findIndexNearestBefore(0.5000001) == 0);
+    SimTK_TEST(states.findIndexNearestBefore(1.299999999) == 0);
+
+    SimTK_TEST(states.findIndexNearestBefore(1.3) == 1);
+    SimTK_TEST(states.findIndexNearestBefore(1.300001) == 1);
+    SimTK_TEST(states.findIndexNearestBefore(3.49999) == 1);
+
+    SimTK_TEST(states.findIndexNearestBefore(4.1) == 3);
+    SimTK_TEST(states.findIndexNearestBefore(4.2) == 3);
+    SimTK_TEST(states.findIndexNearestBefore(153.7) == 3);
+
+    // Nearest after.
+    SimTK_TEST(states.findIndexNearestAfter(-100.0) == 0);
+    SimTK_TEST(states.findIndexNearestAfter(0) == 0);
+    SimTK_TEST(states.findIndexNearestAfter(0.1) == 0);
+    SimTK_TEST(states.findIndexNearestAfter(0.4999999999) == 0);
+    SimTK_TEST(states.findIndexNearestAfter(0.5) == 0);
+
+    SimTK_TEST(states.findIndexNearestAfter(1.3000000000001) == 2);
+    SimTK_TEST(states.findIndexNearestAfter(1.31) == 2);
+    SimTK_TEST(states.findIndexNearestAfter(1.4) == 2);
+    SimTK_TEST(states.findIndexNearestAfter(3.4) == 2);
+    SimTK_TEST(states.findIndexNearestAfter(3.4999999999999) == 2);
+    SimTK_TEST(states.findIndexNearestAfter(3.5) == 2);
+    
+    SimTK_TEST(states.findIndexNearestAfter(4.0001) == 3);
+    SimTK_TEST(states.findIndexNearestAfter(4.1) == 3);
+
+    // Exception if there is not a state before/after the specified time.
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestBefore(0.25),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestBefore(0.0),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestBefore(-1.0),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestBefore(.499999),
+            StatesTrajectory::TimeOutOfRange);
+
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestAfter(4.10000001),
+            StatesTrajectory::TimeOutOfRange);
+    SimTK_TEST_MUST_THROW_EXC(states.findIndexNearestAfter(1050.9),
+            StatesTrajectory::TimeOutOfRange);
+
+    // Simulate the scenario that the tolerance is intended to handle.
```

Comment:
Why is this block here? Why interject states2 at this point when the test case resumes with states? It seems to unnaturally interrupt the flow.